diff --git a/t5x/decoding.py b/t5x/decoding.py
index ecac80e..f308e36 100644
--- a/t5x/decoding.py
+++ b/t5x/decoding.py
@@ -1069,7 +1069,8 @@ def beam_search(inputs: jnp.ndarray,
 
     # If we're not at the max decode length, and the search hasn't terminated,
     # continue looping.
-    return not_at_end & (~search_terminated)  # pytype: disable=bad-return-type  # jax-devicearray
+    #return not_at_end & (~search_terminated)  # pytype: disable=bad-return-type  # jax-devicearray
+    return not_at_end # for debug
 
   def beam_search_loop_body_fn(state: BeamState) -> BeamState:
     """Beam search loop state update function."""
